Add asar support in Atom

zcbenz opened the pull request:
By using [asar](https://github.com/atom/asar) to [package the `app` folder](https://github.com/atom/atom-shell/blob/master/docs/tutorial/application-packaging.md) into a single file,  we should be able to speed up the install time and startup time on computers that run on HDD significantly.

Known issues:
- [x] `index.js` not loading on Windows.
- [x] Can not interpret Less stylesheets because of graceful-fs not compatible with `asar`.
- [x] Program not quitting on Windows.
- [x] Tasks not working.
- [x] `apm` and `atom` commands can not be installed.
- [x] `apm` can not show built-in packages.

zcbenz committed:
:arrow_up: asar@0.2.2

zcbenz committed:
Add generate-asar task

zcbenz committed:
Do not set "cwd" for tasks

The task can be inside an asar archive, which is not allowed to be cwd.

benogle commented:
WIN. :boom:

paulcbetts commented:
This PR is much shorter than I thought it would be, maybe we'll move to asar sooner rather than later

paulcbetts commented on build/tasks/generate-asar-task.coffee:
```diff
@@ -0,0 +1,14 @@
+  grunt.registerTask 'generate-asar', 'Generate asar archive for the app', ->
+    done = @async()
+
+    appDir = grunt.config.get('atom.appDir')
```
Check to make sure this dir exists, otherwise if you run `script/grunt generate-asar` explicitly, you regenerate an empty asar file

paulcbetts commented:
This PR will forever and ever kill all of the long path problems on Windows too, :sparkles:

zcbenz committed:
Print errors that happened before task-bootstrap

zcbenz committed:
Warn empty app dir

Zireael07 commented:
Any progress on the last bullet point?

zcbenz committed:
Copy LICENSE.md to process.resourcesPath

The LICENSE file is needed by mkdeb task, we should move it out of asar
archive.

zcbenz committed:
Install apm to process.resourcesPath

zcbenz committed:
Unpack native module files

It improves performance and makes virus scanners happy

zcbenz commented:
Everything is working now in this branch.

thomasjo commented:
> Everything is working now in this branch.

Awesome! :shipit: :rocket:

zcbenz committed:
Always install commands from process.resourcesPath

Zireael07 commented:
Great!

kevinsawicki commented on src/task.coffee:
```diff
@@ -100,6 +100,11 @@ class Task
       @emit(event, args...) if @childProcess?
+    # Catch the errors that happened before task-bootstrap.
+    @childProcess.stdout.on 'data', (data) ->
+      console.log data.toString()
```
Should these be reported somewhere else? Or just ignored?

zcbenz replied:
This only prints when there is output before the `task-bootstrap` script starts to run, e.g. when there is regression in atom-shell. So as long as atom-shell works as expected, there will be nothing printed.

kevinsawicki commented on build/tasks/build-task.coffee:
```diff
@@ -145,9 +145,9 @@ module.exports = (grunt) ->
     cp 'static', path.join(appDir, 'static')
 
-    cp path.join('apm', 'node_modules', 'atom-package-manager'), path.join(appDir, 'apm'), filter: filterNodeModule
+    cp path.join('apm', 'node_modules', 'atom-package-manager'), path.resolve(appDir, '..', 'apm'), filter: filterNodeModule
```
Since the apm path is changing, does this need to be updated? https://github.com/atom/atom/blob/dd271aae960039b35ce20e2ecfcfc5bc751b42da/src/package-manager.coffee#L132

paulcbetts commented:
![](http://www.oneletterwords.com/assets/Floating-Kitties.gif)

zcbenz committed:
Update path to apm in squirrel-update and package-manager

zcbenz commented:
I just found that apm can not list built-in packages in asar archive, I'll try to fix it in apm.

Zireael07 commented:
There was an apm update in the meantime, I think...

Can't wait for this, should significantly speed up install/update.

zcbenz committed:
:arrow_up: apm@0.153.0

zcbenz committed:
:arrow_up: apm@0.154.0

zcbenz committed:
:arrow_up: apm@0.155.0

kevinsawicki commented:
So changing the `atom.sh` and `apm` paths means that after upgrading, the existing symlinks will be broken right?

So if the person previously ran `Atom > Install Shell Commands` and auth-ed the operation, they will be out of luck until they run it again?

zcbenz commented:
> So if the person previously ran Atom > Install Shell Commands and auth-ed the operation, they will be out of luck until they run it again?

Yeah I believe so. However if they can install the shell commands without privilege I think the commands will be updated silently.

zcbenz committed:
Merge branch 'master' into asar

Conflicts:
	apm/package.json

zcbenz committed:
:arrow_up: apm@0.158.0

zcbenz committed:
InspectorFrontendAPI changed to DevToolsAPI

zcbenz committed:
Fix error when running apm in packages

zcbenz committed:
Merge branch 'master' into asar

zcbenz commented:
There are some random TextEditorComponent specs failing.

Zireael07 commented:
Woot, merged?!

kevinsawicki commented:
Looks like this broke the symbols view so I'm going to revert it for now so it isn't broken on master, https://github.com/atom/symbols-view/issues/89

m1ga commented:
asar will create a tar like file archive, right? How many files would be packed together? So instead of the 11000 files in the windows /app folder you would have only one (136 MB) file?

kevinsawicki commented:
> asar will create a tar like file archive, right? How many files would be packed together? So instead of the 11000 files in the windows /app folder you would have only one (136 MB) file?

Yup, exactly this.

mark-hahn commented:
Does this mean we wouldn't be able to go in and modify code?

On Fri, Apr 3, 2015 at 11:40 AM, Kevin Sawicki <notifications@github.com>
wrote:

> asar will create a tar like file archive, right? How many files would be
> packed together? So instead of the 11000 files in the windows /app folder
> you would have only one (136 MB) file?
>
> Yup, exactly this.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/atom/atom/pull/5382#issuecomment-89385667>.
>

kevinsawicki commented:
> Does this mean we wouldn't be able to go in and modify code?

Yes, dev mode will still load from source though.

mark-hahn commented:
Luckily I'm always in dev mode.  I'm one of those hackers who can't leave
any code untouched.  It's a disease.

On Fri, Apr 3, 2015 at 11:57 AM, Kevin Sawicki <notifications@github.com>
wrote:

> Does this mean we wouldn't be able to go in and modify code?
>
> Yes, dev mode will still load from source though.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/atom/atom/pull/5382#issuecomment-89391655>.
>
